src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt

```diff
@@ -31,28 +32,46 @@ class RsUnresolvedReferenceInspection : RsLocalInspectionTool() {
     override fun buildVisitor(holder: RsProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor =
         object : RsVisitor() {
             override fun visitPath(path: RsPath) {
-                if (path.parent is RsPath) return
-                val (basePath, candidates) = AutoImportFix.findApplicableContext(holder.project, path) ?: return
+                // here we check if this path is resolved
+                // if it is we return
+                // if not than check whether qualifier of this path is resolved
+                // if it is we register problem
+                // if not then we assume that unresolved parent paths of this path
+                // were or will be handled
+                // because all paths will eventually be processed by visitor
+                val (problemPath, candidates) = AutoImportFix.findApplicableContext(holder.project, path)
+                    ?: return
+
                 if (candidates.isEmpty() && ignoreWithoutQuickFix) return
 
-                val referenceName = basePath.identifier?.text
-                // Don't highlight generic parameters
+                path.qualifier
+                    ?.let { AutoImportFix.findApplicableContext(holder.project, it) }
+                    ?.let { return }
+
+                val referenceName = problemPath.identifier?.text
                 val range = TextRange(
-                    0,
-                    path.typeArgumentList?.getPrevNonCommentSibling()?.endOffsetInParent ?: path.textLength
+                    // Don't highlight parent identifiers
+                    problemPath.identifier?.startOffsetInParent ?: 0,
+                    // Don't highlight generic parameters
+                    problemPath.typeArgumentList?.getPrevNonCommentSibling()?.endOffsetInParent
+                        ?: problemPath.textLength
```

Comment:
Let's extract start and end offsets of range into local variables to improve readability and avoid weird formatting